### PR TITLE
feat: add preference option to change MiniSim icon

### PR DESCRIPTION
**Pull request by schrobingus**
Adds an option in the preferences for the user to change the menu bar icon for MiniSim. Addresses #69 (nice).

https://github.com/okwasniewski/MiniSim/assets/72168352/6500bb85-9aeb-48e3-a84e-e2ded421b013

### CODE REVIEW

File: MiniSim/Views/Preferences.swift [hunk 1]
```diff
@@ -11,8 +11,26 @@ import KeyboardShortcuts
...
         Settings.Container(contentWidth: 400) {
+            Settings.Section(title: "Icon:") {
+                Picker("", selection: $menuImageSelected) {
+                    ForEach(0 ..< menuImages.count, id: \.self) { image in
+                        Image(menuImages[image])
+                            .renderingMode(.template)
+                            .tag(image)
+                    }
+                }
+                .pickerStyle(.radioGroup)
```
Comment (okwasniewski) [inline MiniSim/Views/Preferences.swift:27]: Let's use the default picker style (so you can remove this line)

File: MiniSim/Extensions/UserDefaults+Configuration.swift [hunk 2]
```diff
@@ -15,6 +15,7 @@ extension UserDefaults {
         static let isOnboardingFinished = "isOnboardingFinished"
         static let enableiOSSimulators = "enableiOSSimulators"
         static let enableAndroidEmulators = "enableAndroidEmulators"
+        static let menuImage = "menu_icon_1"
```
Comment (okwasniewski) [inline MiniSim/Extensions/UserDefaults+Configuration.swift:18]: Let's create a new file `MenuIcons.swift` 

which would contain enum with all possible options: 

```swift
enum MenuImage: String {
    case iphone = "menu_icon_1"
    case ipad = "menu_icon_2"
    case box = "menu_icon_3"
  }
```

Comment (okwasniewski) [reply MiniSim/Extensions/UserDefaults+Configuration.swift:18]: Also can you rename the assets from menu icon to their actual descriptive name?

File: MiniSim/Views/Preferences.swift [hunk 3]
```diff
@@ -11,8 +11,26 @@ import KeyboardShortcuts
...
 
 struct Preferences: View {
+    var menuImages = [ "menu_icon_1", "menu_icon_2", "menu_icon_3" ]
+    @State var menuImageSelected = 0
+    
     var body: some View {
         Settings.Container(contentWidth: 400) {
+            Settings.Section(title: "Icon:") {
+                Picker("", selection: $menuImageSelected) {
+                    ForEach(0 ..< menuImages.count, id: \.self) { image in
```
Comment (okwasniewski) [inline MiniSim/Views/Preferences.swift:21]: This is an anti-pattern in SwiftUI to pass non static ranges like this. 

When you will create the enum with all images you can do `MenuIcons.allCases`

File: MiniSim/Views/Preferences.swift [hunk 4]
```diff
@@ -11,8 +11,26 @@ import KeyboardShortcuts
 import LaunchAtLogin
 
 struct Preferences: View {
+    var menuImages = [ "menu_icon_1", "menu_icon_2", "menu_icon_3" ]
+    @State var menuImageSelected = 0
```
Comment (okwasniewski) [inline MiniSim/Views/Preferences.swift:15]: I think you can use the `@AppStorage` property wrapper to save directly to user defaults 👍🏻

File: MiniSim/Views/Preferences.swift [hunk 5]
```diff
@@ -31,6 +49,13 @@ struct Preferences: View {
             }
         }
         .frame(minWidth: 650, minHeight: 450)
+        .onAppear {
```
Comment (okwasniewski) [inline MiniSim/Views/Preferences.swift:52]: If you will use `@AppStorage` this also won't be needed

File: MiniSim/MiniSim.swift [hunk 6]
```diff
@@ -14,9 +14,12 @@ import UserNotifications
 class MiniSim: NSObject {
     private var menu: Menu!
     
-    @objc let statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
+    @objc var statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
     
     private var isOnboardingFinishedObserver: NSKeyValueObservation?
+    private var menuImageObserver: NSKeyValueObservation?
+    
+    private var menuImagesObserved: [String] = []
```
Comment (okwasniewski) [inline MiniSim/MiniSim.swift:22]: Why do we need to store those images? This looks unnecessary

File: MiniSim/MiniSim.swift [hunk 7]
```diff
@@ -110,8 +117,13 @@ class MiniSim: NSObject {
     private func setMenuImage() {
         if let button = statusItem.button {
             button.toolTip = "MiniSim"
-            let itemImage = NSImage(named: "menu_icon")
-            itemImage?.size = NSSize(width: 9, height: 16)
+            let itemImage = NSImage(named: UserDefaults.standard.menuImage)
+            if (!menuImagesObserved.contains(UserDefaults.standard.menuImage)) {
+                itemImage?.size = NSSize(
```
Comment (okwasniewski) [inline MiniSim/MiniSim.swift:122]: You can set the size without this if check as we have the `?.size` optional check

**Review by okwasniewski — CHANGES_REQUESTED** [review]
Thanks for those changes! I've left few comments

**Comment by BrentBoyMeBob** [comment]
I'll be on it!

File: MiniSim/MiniSim.swift (hunk 6 above)
Comment (BrentBoyMeBob) [reply MiniSim/MiniSim.swift:22]: menuImagesObserved is used because without a conditional that uses it (line 121), when you select an icon again that was already used before, hence already resized and padded, it'll get smaller and smaller. This prevents this from happening. There could be a cleaner approach to this, but I'm not sure.

File: MiniSim/Views/Preferences.swift (hunk 4 above)
Comment (BrentBoyMeBob) [reply MiniSim/Views/Preferences.swift:15]: From my tests, AppStorage can be used, but the Picker's appearance won't update without State. I can't add both properties in one variable.

**Comment by BrentBoyMeBob** [comment]
I've made the changes suggested, albeit I do have a concern regarding @AppStorage's usage. I've resolved the suggestions that have been applied. :)

File: MiniSim/Extensions/UserDefaults+Configuration.swift [hunk 8]
```diff
@@ -27,6 +28,11 @@ extension UserDefaults {
         set { set(newValue, forKey: Keys.isOnboardingFinished) }
     }
     
+    @objc dynamic public var menuImage: String {
```
Comment (okwasniewski) [inline MiniSim/Extensions/UserDefaults+Configuration.swift:31]: Can you make it return `MenuImage` enum that you've added?

And also as the default let's use `MenuImage.iphone`

File: MiniSim/MiniSim.swift [hunk 9]
```diff
@@ -14,9 +14,10 @@ import UserNotifications
 class MiniSim: NSObject {
     private var menu: Menu!
     
-    @objc let statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
+    @objc var statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
```
Comment (okwasniewski) [inline MiniSim/MiniSim.swift:17]: Why is this change needed?

File: MiniSim/MiniSim.swift [hunk 10]
```diff
@@ -106,13 +112,21 @@ class MiniSim: NSObject {
             UserDefaults.Keys.enableiOSSimulators: true
         ])
     }
+   
+    private func configureMenuImages() {
```
Comment (okwasniewski) [inline MiniSim/MiniSim.swift:116]: let's move this logic to the `MenuImage` enum: 

```swift
enum MenuImage: String, CaseIterable {
    case iphone = "iphone"
    case ipad = "ipad"
    case box = "box"
    
    var image: NSImage? {
        guard let itemImage = NSImage(systemSymbolName: self.rawValue, accessibilityDescription: self.rawValue) else {
            return nil
        }
        itemImage.size = NSSize(
            width:  (itemImage.size.width)  * 0.78,
            height: (itemImage.size.height) * 0.78)
        itemImage.isTemplate = true
        return itemImage
    }
}
```

Comment (okwasniewski) [reply MiniSim/MiniSim.swift:116]: Then you can call `MenuImage.iphone.image`

File: MiniSim/Views/Preferences.swift [hunk 11]
```diff
@@ -11,8 +11,24 @@ import KeyboardShortcuts
 import LaunchAtLogin
 
 struct Preferences: View {
+    @State var menuImageSelected: String = "iphone"
```
Comment (okwasniewski) [inline MiniSim/Views/Preferences.swift:14]: I'm using `@AppStorage` in `MiniSim/Views/Onboarding/SetupView.swift` and it works great with Picker so I'm not sure whats the issue here. Can you check out that file?

File: MiniSim/MiniSim.swift [hunk 12]
```diff
@@ -106,13 +112,21 @@ class MiniSim: NSObject {
...
+        MenuImage.allCases.forEach { image in
+            let itemImage = NSImage(imageLiteralResourceName: image.rawValue)
```
Comment (okwasniewski) [inline MiniSim/MiniSim.swift:118]: According to docs this initializer of NSImage shouldn't be used directly

File: MiniSim/Views/Preferences.swift (hunk 4 above)
Comment (okwasniewski) [reply MiniSim/Views/Preferences.swift]: I've used `AppStorage` for Picker in this file: `MiniSim/Views/Onboarding/SetupView.swift` - check it out

File: MiniSim/Extensions/UserDefaults+Configuration.swift (hunk 8 above)
Comment (BrentBoyMeBob) [reply MiniSim/Extensions/UserDefaults+Configuration.swift:31]: Objective-C compliance will have to be removed from menuImage, but I was able to rework the observer to that.

File: MiniSim/MiniSim.swift (hunk 9 above)
Comment (BrentBoyMeBob) [reply MiniSim/MiniSim.swift:17]: Thought the immutability affected image changes but I was incorrect, that's on me.

File: MiniSim/MiniSim.swift (hunk 10 above)
Comment (BrentBoyMeBob) [reply MiniSim/MiniSim.swift:116]: configureMenuImages is only ran once because the image gets smaller and smaller every single time the function is reran, NSImage remembers the size. Pardon me if I misunderstand, but this would supposedly assume that this variable would be called multiple times?

Comment (okwasniewski) [reply MiniSim/MiniSim.swift:116]: Image was getting smaller and smaller because you was mutating the button image in this case each item would have its own image

File: MiniSim/Views/Preferences.swift (hunk 11 above)
Comment (BrentBoyMeBob) [reply MiniSim/Views/Preferences.swift:14]: @okwasniewski Sorry for the long delay, I've been a bit busy lately. SetupView works because `enableiOSSimulators` and `enableAndroidSimulators` are changed before the menu setup is called. With `menuImageSelected`, it's a bit different, because the menu needs to change on the fly, as a `@State` would. I have seen examples of `@AppStorage` acting this way, but it isn't recurring for me.

For example, if I call `enableAndroidSimulators` in this way:
```swift
struct Preferences: View {
    ...
    @AppStorage(UserDefaults.Keys.enableAndroidEmulators, store: .standard) var enableAndroidEmulators: Bool = true

    var body: some View {
        Settings.Container(contentWidth: 400) {
                ...
                Settings.Section(title: "Enable Android Emulators:") {
                        Toggle(isOn: $enableAndroidEmulators) {
                        Text("Enable Android Emulators")
                }
                ...
 ```
 it doesn't mutate whilst it's open, because at least in my case, `@AppStorage` doesn't act like State.
